### DISCUSSION TITLE
LDM Notes for February 28th, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5853] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-02-28.md

## Agenda

1. Ref fields
    1. Encoding strategy
    2. Keywords vs Attributes
    3. Breaking existing lifetime rules

**Comment by agocke** [comment]
An alternate syntax approach would be to encode generic lifetimes directly. We would need annotations for ref variables and ref structs. They could look like a different kind of generic:

```C#
ref struct Span<T>{a}
{
    readonly ref{a} T _field;
    int _length;
}
```

In this case, the `Span` type has a lifetime `a` and so does `_field`, which means that `_field` can't live beyond `a`. `a` is a generic lifetime, because we don't know how the Span is going to be created yet. If it's created like this:

```C#
Span<int> s = new int[10];
```

then the lifetime of a heap-allocated array is global, so `s` has a global lifetime. If it's create like this:

```C#
Span<int> s = stackalloc int[10];
```

then the lifetime of the span is the current method, since that's the lifetime of a `stackalloc`. We can then extend the concept to `ref` parameters. The following code is illegal today:

```C#
    public ref int M(ref int x) {
        int y = 0;
        return ref M2(ref x, ref y); // error
    }
    public ref int M2(ref int x, ref int y)
    {
        return ref x;
    }
```
because the language doesn't know that the return lifetime of `M2` is only dependent on the `x` parameter, not the `y` parameter. With explicit annotation, we could tell it that:

```C#
    public ref int M(ref int x) {
        int y = 0;
        return ref M2(ref x, ref y);
    }
    public ref{a} int M2{a, b}(ref{a} int x, ref{b} int y)
    {
        return ref x;
    }
```

Usually you probably wouldn't have to tell the compiler anything, since the default lifetimes mostly make sense.

We could also probably use this to allow ref structs in generics, by saying that generic can be substituted with a ref struct if it has an explicit lifetime.

```C#
void M<T>{a}(T{a} t)
{
   ...
}
```

Since `T` has a lifetime, it's not implicitly global scoped, so it wouldn't be safe to box.

**Reply by 333fred** [reply]
And just like that, Rust is born.
